### PR TITLE
Replace QTimer in GUI by QtToTrioJobScheduler.submit_delayed_job

### LINKED ISSUE #1740: Avoid QTimer in the GUI
see #1739

We already avoid `QTimer` as much as possible given it lead to complex spaghetti code and polling. Two more reasons to avoid it:
- The GUI tests are greatly slowed down when waiting for something that is triggered by a `QTimer`
- The GUI tests can even end up in hard to reproduce timeout (typically with a timeout of 3s and a `QTimer` triggered every second), see for instance https://dev.azure.com/Scille/parsec/_build/results?buildId=6032&view=logs&j=5ba45810-c0e9-5996-abe7-b2cb0c9b6baa&t=b18150ab-3a6d-5272-0b08-a0dc06a2acd4

The place we use `QTimer`:
```
$ rg QTimer parsec/core/gui
parsec/core/gui/workspaces_widget.py
6:from PyQt5.QtCore import pyqtSignal, QTimer, Qt
240:        self.reset_timer = QTimer()

parsec/core/gui/files_widget.py
11:from PyQt5.QtCore import Qt, pyqtSignal, QTimer
305:        self.update_timer = QTimer()

parsec/core/gui/app.py
14:from PyQt5.QtCore import QTimer, Qt
215:        # QTimer wakes up the event loop periodically which allows us to close
217:        timer = QTimer()
222:            diagnose_timer = QTimer()

parsec/core/gui/central_widget.py
4:from PyQt5.QtCore import pyqtSignal, QTimer
81:        self.organization_stats_timer = QTimer()

parsec/core/gui/workspace_sharing_widget.py
5:from PyQt5.QtCore import QCoreApplication, pyqtSignal, QEvent, QTimer
122:        self.status_timer = QTimer()
```

> parsec/core/gui/app.py

From what I recall, this one is hard to remove given it aims at forcing
Qt loop to do something so we can close the main window event if it is in the background...
We may want to put more investigation into the reason of this (I'd like to see
more documentation about this in the code given it's quiet exotic ^^), but
no need to remove `QTimer` from here I think.

> parsec/core/gui/workspaces_widget.py

This is the bad boy that cause all the GUI test to slow down and timeout, should
definitively remove it and replace it by core event handling.

> parsec/core/gui/files_widget.py

Pretty sure this one cause the same kind of issues that `workspaces_widget.py`

> parsec/core/gui/central_widget.py

This one is here only to send a `organization_stats_timer` command every 5s to
the backend...
Seems pretty wasteful, I think we could trigger the stat command when the window
is created.
I would even remove this entirely and add a this information in the logout menu
(so that the command is send only when the menu is opened)

> parsec/core/gui/workspace_sharing_widget.py

This could be replaced by events from the sync monitor

### PR DESCRIPTION
**Pull request by touilleMan**
The non-hacky brother of #1739

fix #1740

### CODE REVIEW

File: parsec/core/gui/app.py [hunk 1]
```diff
@@ -215,8 +215,8 @@ def kill_window(*args):
         # QTimer wakes up the event loop periodically which allows us to close
         # the window even when it is in background.
         timer = QTimer()
-        timer.start(1000 if diagnose else 400)
-        timer.timeout.connect(kill_window if diagnose else lambda: None)
+        timer.start(400)
+        timer.timeout.connect(lambda: None)
```
Comment (vxgmichel) [inline parsec/core/gui/app.py:219]: Any reason for removing `diagnose`?

File: parsec/core/gui/files_widget.py [hunk 2]
```diff
@@ -628,13 +626,15 @@ def item_activated(self, file_type, file_name):
         elif file_type == FileType.Folder:
             self.load(self.current_directory / file_name)
 
-    def reload(self):
-        self.load(self.current_directory)
+    def schedule_reload(self):
+        self.load(self.current_directory, delay=self.RELOAD_FILES_LIST_DELAY)
```
Comment (vxgmichel) [inline parsec/core/gui/files_widget.py:630]: Renaming to `schedule_reload` is no longer justified.

File: parsec/core/gui/trio_thread.py [hunk 3]
```diff
@@ -200,6 +205,62 @@ def _run_job(self, job, *args, sync=False):
...
+            nonlocal qt_on_success, qt_on_error, args, kwargs
+            submitted = trio.current_time()
+
+            # Only modify `_scheduled_jobs_args` from the trio
+            # thread to avoid concurrent acces with the Qt thread
+            self._scheduled_jobs_args[throttling_id] = (qt_on_success, qt_on_error, args, kwargs)
+            task_status.started()
+
+            last_executed = self._scheduled_jobs_last_executed.get(throttling_id, 0)
+            await trio.sleep_until(last_executed + delay)
```
Comment (vxgmichel) [inline parsec/core/gui/trio_thread.py:233]: Maybe add a comment to show that sleeping is only needed if a similar job ran in the last `delay` seconds.

File: parsec/core/gui/trio_thread.py [hunk 4]
```diff
@@ -200,6 +205,62 @@ def _run_job(self, job, *args, sync=False):
...
+
+            # Make sure fn has not been executed while we were waiting
+            job_args = self._scheduled_jobs_args.pop(throttling_id, None)
+            last_executed = self._scheduled_jobs_last_executed.get(throttling_id, 0)
+            if not job_args or last_executed > submitted:
+                return
```
Comment (vxgmichel) [inline parsec/core/gui/trio_thread.py:239]: There's a bug here: `job_args` has been popped but is discarded if `last_executed > submitted`, which means the other scheduled job will have nothing to pop from `_scheduled_jobs_args`

Comment (vxgmichel) [reply parsec/core/gui/trio_thread.py:239]: Also I'm not a big fan of `submitted` as it adds an extra timestamp (and extra race conditions).

Instead maybe use `last_executed + delay > trio.current_time()` which doesn't require the extra variable.

**Review by vxgmichel — CHANGES_REQUESTED** [review]
A few comments, LGTM otherwise :+1:

File: parsec/core/gui/app.py (hunk 1 above)
Comment (touilleMan) [reply parsec/core/gui/app.py:219]: diagnose is already handled in the next lines, I didn't see the need to add it here (we could also remove the `if diagnose` stuff and keep this timer multi-purposed)

File: parsec/core/gui/trio_thread.py (hunk 4 above)
Comment (touilleMan) [reply parsec/core/gui/trio_thread.py:245]: Good catch !

> Instead maybe use last_executed + delay > trio.current_time() which doesn't require the extra variable.

:+1:

File: parsec/core/gui/trio_thread.py (hunk 3 above)
Comment (touilleMan) [reply parsec/core/gui/trio_thread.py:236]: yup, I love comments !

File: parsec/core/gui/files_widget.py (hunk 2 above)
Comment (touilleMan) [reply parsec/core/gui/files_widget.py]: :+1:

**Review by vxgmichel — APPROVED** [review]
:+1:

**Comment by vxgmichel** [comment]
Aouch: https://dev.azure.com/Scille/parsec/_build/results?buildId=6069&view=logs&j=5ba45810-c0e9-5996-abe7-b2cb0c9b6baa&t=b18150ab-3a6d-5272-0b08-a0dc06a2acd4&l=54

**Comment by vxgmichel** [comment]
@touilleMan GUI tests should now be consistent \o/ (at least `pytest --rungui -m gui --count 10` runs on my machine).

Could you review the changes I made?

File: tests/core/gui/test_main_window.py [hunk 5]
```diff
@@ -110,6 +110,8 @@ def _workspace_button_ready():
         assert w_w.layout_workspaces.count() == 1
         wk_button = w_w.layout_workspaces.itemAt(0).widget()
         assert not isinstance(wk_button, QtWidgets.QLabel)
+        # Important: this is necessary for the incoming click to reliably work
```
Comment (touilleMan) [inline tests/core/gui/test_main_window.py:113]: Not sure why this is important: this check box enable workspace mounting, however I don't think we test the mountpoint in GUI test
(on top of that I don't remember the rule for a new workspace to be mounted or not...)

File: parsec/core/gui/files_widget.py [hunk 6]
```diff
@@ -635,20 +636,31 @@ def item_activated(self, file_type, file_name):
         elif file_type == FileType.Folder:
             self.load(self.current_directory / file_name)
 
-    def reload(self):
-        self.load(self.current_directory, throttle_delay=self.RELOAD_FILES_LIST_THROTTLE_DELAY)
+    def reload(self, default_selection=None, delay=None):
```
Comment (touilleMan) [inline parsec/core/gui/files_widget.py:639]:
```python
def reload(self, default_selection=None, delay=RELOAD_FILES_LIST_THROTTLE_DELAY):
```
?

**Review by touilleMan — COMMENTED** [review]
lgtm (but can't auto-approve my own PR 😢 )

File: tests/core/gui/test_files.py [hunk 7]
```diff
@@ -107,14 +107,7 @@ def _do_selection():
...
         async def reset_selection(self):
-            for r in f_w.table_files.selectedRanges():
-                top = r.topRow()
-                bottom = r.bottomRow()
-                await aqtbot.run(
-                    f_w.table_files.setRangeSelected,
-                    QtWidgets.QTableWidgetSelectionRange(top, 0, bottom, 0),
-                    False,
-                )
+            await aqtbot.run(f_w.table_files.reset)
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply tests/core/gui/test_files.py:110]: Both!

File: tests/core/gui/test_main_window.py (hunk 5 above)
Comment (vxgmichel) [reply tests/core/gui/test_main_window.py:113]: It is important as left click is disabled if the check box is checked (i.e the workspace is not mounted). This was the reason why some tests were inconsistent, when the click happened to be sent before the the check box was enabled.

File: tests/core/gui/test_files.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vxgmichel) [reply tests/core/gui/test_files.py:110]: I didn't look much into it but I noticed some selection could remain in some cases after a call to `reset_selection`. Simply using the dedicated qt call fixed the issue.

File: parsec/core/gui/files_widget.py (hunk 6 above)
Comment (vxgmichel) [reply parsec/core/gui/files_widget.py]: d1812a2

**Pull request merged**